### PR TITLE
Make `ForEach-Object` faster for its commonly used scenarios

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

**NOTE:** mark the PR as WIP because the rewriting causes some issues to stepping during debugging. Looking into that.

A follow-up PR for #10047
Make `ForEach-Object` faster for its commonly used scenarios by rewriting the command into a filter-like script block execution in the pipeline.
For the follow `ForEach-Object` usages, they will be rewritten to `... | . { process { ... } } | ...` when constructing the pipeline.

```powershell
... | % { ... } | ...
... | % -process { ... } | ...
... | % -process:{ ... } | ...
```

By doing this rewriting, we are able to eliminate all unnecessary overheads from:
1. parameter binding
2. invoking a script block with `ScriptBlock.InvokeUsingCmdlet` for every incoming object, which has to do the setting-up/tearing-down every time executing the script block.

This PR targets to improve the performance of `ForEach-Object` for the most common usages. It's not a goal to cover all possible scenarios, such as `% -pro:{...}` or `% { ... } { ... } { ... }`

| Script | Before | After |
|-|-|-|
| `1..100kb \| % { }` | 290ms | 37ms |
|`1..100kb \| . { process {} }`| 37ms | 37ms |
| `foreach ($i in 1..100kb) { }` | 28ms | 28ms |

This optimization will be turned off automatically in the following cases:
1. when debugger is enabled, namely when any breakpoint is set in the session. This is to make the debugger works as expected in case a breakpoint is set on `ForEach-Object`, e.g `Set-PSBreakpoint -Command ForEach-Object`.
2. when 'ConstrainedLanguageMode' has been used for the current Runspace. The language mode transition is tricky. It's better to stick with the same old code path for safety.

## PR Context

https://github.com/PowerShell/PowerShell/issues/9731#issuecomment-496721086
Draft PR: #10153 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [x] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 1]
```diff
@@ -311,6 +315,89 @@ internal static class PipelineOps
...
+            const string ForEachObject_ProcessParam = "Process";
+
+            if (commandProcessor.CommandInfo is CmdletInfo cmdlet && cmdlet.ImplementingType == typeof(ForEachObjectCommand))
+            {
+                int indexAdvanceOffset = 0;
+                int cmdElementsLength = commandElements.Length;
+                ScriptBlock processScriptBlock = null;
+
+                if (commandIndex == cmdElementsLength - 1)
+                {
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: Please add comments here and below what cmdlet form is each branch addressed.

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 2]
```diff
@@ -311,6 +315,89 @@ internal static class PipelineOps
...
+                        processScriptBlock = nextElement.ArgumentValue as ScriptBlock;
+                        indexAdvanceOffset = 2;
+                    }
+                }
+
+                if (processScriptBlock != null && processScriptBlock.Ast is ScriptBlockAst sbAst)
+                {
+                    if (!sbAst.IsConfiguration && sbAst.ParamBlock == null && sbAst.BeginBlock == null &&
+                        sbAst.ProcessBlock == null && sbAst.DynamicParamBlock == null && sbAst.EndBlock != null &&
+                        sbAst.EndBlock.Unnamed)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: Do you consider moving this in parse time? Maybe we could have a special ast type? I am thinking about Where-Object where we could use the same way to optimize performance.

**Review by rjmholt — APPROVED** [review]
Naturally I'd be interested to know if there's some other body or context I could execute the optimised and unoptimised version in that isn't debugging where the behaviour would differ, but I can't think of one

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: > it would require adding new elements to the existing script block, and that will cause issue to the accuracy of extent, debugging, and more, which would be too much magic to deal with.

I see currently we skip optimization if we debug so it is not problem. 
If we look https://learn-powershell.net/2015/02/20/a-look-at-filtering-performance-in-powershell/ we can see that filter function has best performance. We could convert scriptblock to a filter function. This conversion does not look much magic. If we could still add parameter binding optimization we could get great performance win for most of common scenarios.

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 3]
```diff
@@ -311,6 +315,104 @@ internal static class PipelineOps
...
+                sbAst.Parent?.SetParent(newScriptBlockAst);
+
+                return new ScriptBlock(newScriptBlockAst, isFilter: false);
+            };
+
+        private static bool TryRewriteForEachObjectCommand(
+            ExecutionContext context,
+            SessionStateInternal commandSessionState,
+            CommandParameterInternal[] commandElements,
+            ref CommandProcessorBase commandProcessor,
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: Why do we need `ref` for the class parameter?

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: Perhaps it is more easy at parse time as I already mentioned, I did something like while did experiments with class getters/setters. (We could enhance Parser/Compiler to support this and it could become a common approach for dynamic optimizations. It might be similar to what CoreCLR does today by postponing compilation and optimizations.)

Ah, I found that we have #9941 tracking issue for Where-Object.

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: CommandProcessorBase is a reference type and I'd expect that it is passed by reference by default.

**Comment by KirkMunro** [comment]
I always assumed optimizations like this would be handled in an `AstVisitor` implementation, hoping such optimizations could be performed while retaining the extent in the original scripts so that neither the script author nor the debugger is aware of them while code just runs faster. This PR seems to indicate that my assumption is incorrect.

My naive questions about this: Is it simply not possible to perform optimizations such as this one transparently to the user and the debugger by doing the work in the AST itself? Or is it possible, but an approach like that is considered undesirable for other reasons?

**Review by PaulHigin — APPROVED** [review]

**Comment by daxian-dbw** [comment]
> Is it simply not possible to perform optimizations such as this one transparently to the user and the debugger by doing the work in the AST itself?

The original AST should be preserved exactly as is because it should reflect exactly what the user typed on the console or wrote in the file.
If your thought is to generate another AST via a visitor before compiling, and use that one for compilation, then there is no guarantee the extent will be exactly the same and it will be transparent to the debugger.
Take this change as an example, once `foreach-object { ... }` is replaced by `. { process { ... } }` in the generated AST, you will have problem on the extent of the `process {` part, because that's new elements added the generated AST. Also, there won't be a `CommandProcessor` representing the `ForEach-Object` in the pipeline, and thus the debugger will break at the open curly instead of `ForEach-Object` for `Set-PSBreakpoint -Command Foreach-Object`.

For this specific change, I think there is something we can try if we really want to make it work with debugger: when rewriting the pipeline, instead of creating a command processor with the script block, we can construct a `FunctionInfo` with the name `ForEach-Object` and make its extent the same as the `ForEach-Object` extent. That might work, and probably worth explore as a follow-up PR.

**Comment by rjmholt** [comment]
@KirkMunro 

> Is it simply not possible to perform optimizations such as this one transparently to the user and the debugger by doing the work in the AST itself? Or is it possible, but an approach like that is considered undesirable for other reasons

It's definitely possible. There's a good example in `DynamicKeywordStatementAst`:

https://github.com/PowerShell/PowerShell/blob/704ee90d5cd2909ebc98a8d624359eed5d1b1cbc/src/System.Management.Automation/engine/parser/ast.cs#L6817-L7027

But to my mind, the questions are those of layering:

- How faithfully should the AST represent the program as it was given to the parser?
- How much should the AST know about what consumes it?
- To what extent (heh) should I be able to reconstruct a program from its AST?
- If we add syntactic sugar, does the parser own it, the AST or the compiler?

These problems (/restatements of the same problem) are inconsistently addressed in PowerShell:

- [The parser doesn't record where traps occur in a block at all, so the AST is created with some arguably important information missing](https://github.com/PowerShell/PowerShell/blob/704ee90d5cd2909ebc98a8d624359eed5d1b1cbc/src/System.Management.Automation/engine/parser/Parser.cs#L1858-L1908)
- [Dynamic keyword invocations get their own AST, but they do their own AST transformation to be compiled](https://github.com/PowerShell/PowerShell/blob/704ee90d5cd2909ebc98a8d624359eed5d1b1cbc/src/System.Management.Automation/engine/parser/Compiler.cs#L2978-L2986)
- The compiler's `*Ops` libraries have some syntactic sugar transformations in them, like [for job control operators (`Get-Command &`)](https://github.com/PowerShell/PowerShell/blob/704ee90d5cd2909ebc98a8d624359eed5d1b1cbc/src/System.Management.Automation/engine/parser/Compiler.cs#L3092-L3098) and here with the `ForEach-Object` now

If we were writing it all over today, we might have a syntactic sugar phase/layer between the parser and the compiler implemented as an AstVisitor that takes in one AST and spits out the transformed one, although the immutability of the AST (which I still think is a very good thing) makes that a bit more expensive. That's certainly the way other statically compiled. languages have gone.

But I do feel ambivalent about syntactic sugar (in the strict sense of a language-level AST-to-AST transformation, where one syntax is a macro for another) in an interpreted language. I'm not sure about Python, Ruby and friends, but both C# and Java seem to have a policy against it as far as I can tell; they go direct to bytecode and leave the optimisations to the runtime.

For PowerShell though, I think (1) we want to encourage good practice with syntax, (2) it's good to try and find performance wherever we can while still finding time for other work.

So basically, I don't think we have a good, consistent answer for how something like this should be implemented, but this implementation is one of the better ones in terms of both how it's done and where it sits.

**Comment by KirkMunro** [comment]
Those answers are very, very helpful. Thank you @daxian-dbw and @rjmholt for taking the time to provide that information.

**Comment by iSazonov** [comment]
I like the optimization because it works well for _all_ scenarios.
Although it’s obvious that there could be optimizations that can work well in a particular situation. That's why the question of Ast optimizations arises.

For example, for a simple management script in a task scheduler, an optimization is not to do any optimization at all since the interpreter will be the fastest.
For data manning script (parse and analyze large log file(s)), we could say in advance `pwsh -Optimization Full` to apply all optimizations _before_ run. In the scenario generating optimized Ast tree could bring significant acceleration. Currently single optimization we do for such scenarios - switch from interpreter to compiler for long running script blocks.

> they go direct to bytecode and leave the optimizations to the runtime.

Since we have new modern runtime we could review if we can generate code so that it can optimize by the runtime better.

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: Thanks for clarify! Look like a trick because it is unusual for reference types. Protection comment would be good.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: OK, will add a comment here.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: I thought about what the comment should be, but the more I want to explain it clear the more I realize it become a comment that tries to explain a C# language feature ...
This is how `ref` parameter is used in C#, it passes in the variable address, and using the `ref` parameter in the method will be like directly operating on the variable from the caller method. This is essentially how local functions can directly change the variables in the enclosing methods.

Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: That's true, but given most analysis tools typically issue a warning for using `ref` with a reference type, a comment is probably advisable so that it's not accidentally removed. Even if that comment is "yes this is intended, don't change it" :)

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: I searched the code base, there are many places `ref` is used for a reference type parameter, such as `Exception`. `Interlock.Exchange (ref object location1, object value)` also uses `ref` for reference type.
There is no difference between `ref CommandProcessorBase commandProcessor` and `ref int commandIndex` because both are aimed to directly mutate the variables in the caller method.
Personally, I don't think a comment is needed here.

Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs]: #Closed

**Review by adityapatwardhan — APPROVED** [review]

**Comment by KirkMunro** [comment]
@daxian-dbw This PR introduces a breaking change. :boom:

Step 1: Create a file with the following contents:

```PowerShell
1..2 | ForEach-Object{

    Write-Host('MyInvocation in ForEach-Object: ' + $MyInvocation.MyCommand.Name)

}
```

Step 2: Dot-source or invoke that file with the call operator.

Before this change (using preview 3):
`$MyInvocation.MyCommand.Name` results in the name of the file.

After this change (using a custom build with the latest bits, pulled just now):
`$MyInvocation.MyCommand.Name` is `$null`.

**Comment by daxian-dbw** [comment]
Interesting, I knew there might be cases that I didn't catch :) I will be working on a fix.

I guess we also need an answer to this question:
> why does the behaviour differ between Where-Object and ForEach-Object? They are both cmdlets, and I don't think it's unreasonable for users to expect a consistent experience between these two commands in terms of how their scriptblock implementations behave.

**Comment by daxian-dbw** [comment]
#10485 was submitted to revert this PR given that it introduced a breaking change to the value of `$MyInvocation` within the script block argument.
